lib/smart_proxy_dns_powerdns/backend/rest.rb

```diff
@@ -0,0 +1,101 @@
+require 'json'
+require 'net/http'
+require 'resolv'
+
+module Proxy::Dns::Powerdns::Backend
+  class Rest < ::Proxy::Dns::Powerdns::Record
+
+    attr_reader :url, :api_key
+
+    def initialize(a_server = nil, a_ttl = nil)
+      @url = Proxy::Dns::Powerdns::Plugin.settings.powerdns_rest_url
+      @api_key = Proxy::Dns::Powerdns::Plugin.settings.powerdns_rest_api_key
+
+      super(a_server, a_ttl)
+    end
+
+    def get_zone name
+      fqdn = Resolv::DNS::Name.create(name)
+      fqdn = Resolv::DNS::Name.create(name + '.') unless fqdn.absolute?
+      uri = URI("#{url}/zones")
+
+      domain = nil
+
+      Net::HTTP.start(uri.host, uri.port) do |http|
+        request = Net::HTTP::Get.new uri
+        request['X-API-Key'] = @api_key
+        response = http.request request
+        zones = JSON.parse(response.body) rescue []
+
+        domain = zones.select { |zone|
+          domain = Resolv::DNS::Name.create(zone['name'])
+          domain == fqdn or fqdn.subdomain_of?(domain)
+        }.max_by { |zone| zone['name'].length }
+      end
+
+      raise Proxy::Dns::Error, "Unable to determine zone. Zone must exist in PowerDNS." unless domain
+
+      domain
+    end
+
+    def create_record domain_id, name, type, content
+      content += '.' if type == 'PTR'
+      rrset = {
+        :name => name + '.',
+        :type => type,
+        :ttl => @ttl.to_i,
+        :changetype => :REPLACE,
+        :records => [
+          {
+            :content => content,
+            :disabled => false
+          }
+        ]
+      }
+
+      patch_records domain_id, rrset
+    end
+
+    def delete_record domain_id, name, type
+      rrset = {
+        :name => name + '.',
+        :type => type,
+        :changetype => :DELETE,
+        :records => []
+      }
+
+      patch_records domain_id, rrset
+    end
+
+    def rectify_zone domain
+      # Not needed by the API
+      true
+    end
+
+    private
+
+    def patch_records domain_id, rrset
+      uri = URI("#{url}/zones/#{domain_id}")
+
+      data = { :rrsets => [rrset] }
+
+      Net::HTTP.start(uri.host, uri.port) do |http|
+        request = Net::HTTP::Patch.new uri
+        request['X-API-Key'] = @api_key
+        request['Content-Type'] = 'application/json'
+        request.body = data.to_json
+        response = http.request request
+        unless response.is_a?(Net::HTTPSuccess)
+          begin
+            content = JSON.parse(response.body)
+          rescue
+            raise Proxy::Dns::Error.new("Failed to patch records")
```

Comment:
Probably worth mentioning that the response couldn't be parsed, maybe even log the contents of response on debug?